File: tennis/tennis.rb
class Tennis
	def initialize
		@player1 = 0
		@player2 = 0
	end
	def description
		in_deuce? ? deuce_stage : normal_stage
	end

	def player1_win
		@player1 += 1
	end
	def player2_win
		@player2 += 1
	end

	private
	def scoring score
		["love", "fifteen", "thirty", "forty"][score]
	end

	def in_deuce?
		@player1 >= 3 && @player2 >= 3
	end

	def normal_stage
		if @player1 == 4
			desc = "player1 wins"
		elsif @player2 == 4
			desc = "player2 wins"
		else
			player1_scoring = scoring @player1	
			player2_scoring = scoring @player2
			desc = "#{player1_scoring} #{player2_scoring}"
		end
		desc
	end

	def deuce_stage
		if @player1 == @player2
			desc = "deuce"
		end
		case @player1 - @player2
		when 1
			desc = "player1 advantage"
		when 2
			desc = "player1 wins"
		when -1
			desc = "player2 advantage"
		when -2
			desc = "player2 wins"
		end
		desc
	end
end

File: tennis/java/src/Tennis.java
public class Tennis {
    private final static String[] SCORE_WORD = {"Love", "Fifteen", "Thirty", "Forty"};

    private int player1Score = 0;
    private int player2Score = 0;

    public String score() {
        if (isGameEnd())
            return gameResult();

        if (isInDeuce())
            return scoreInDeuce();

        return normalScore();
    }

    private String normalScore() {
        String scoreDescription;
        if (isDraw())
            scoreDescription = String.format("%s All", SCORE_WORD[player1Score]);
        else
            scoreDescription = String.format("%s %s", SCORE_WORD[player1Score], SCORE_WORD[player2Score]);
        return scoreDescription;
    }

    private String scoreInDeuce() {
        String scoreDescription;
        if (isDraw())
            scoreDescription = "Deuce";
        else if (player1Score > player2Score){
                scoreDescription = "Player1 Advance";
        }
        else{
                scoreDescription = "Player2 Advance";
        }
        return scoreDescription;
    }

    private boolean isGameEnd() {
        return Math.abs(player1Score - player2Score) > 1 && (player1Score >= 4 || player2Score >=4);
    }

    private String gameResult() {
        return player1Score > player2Score ? "Player1 Wins" : "Player2 Wins";
    }

    private boolean isDraw() {
        return player1Score == player2Score;
    }

    private boolean isInDeuce() {
        return player1Score >= 3 && player2Score >= 3;
    }

    public void player1_scores() {
        player1Score += 1;
    }

    public void player2_scores() {
        player2Score += 1;
    }
}
